src/ophyd_async/core/_detector.py

```diff
@@ -193,9 +197,12 @@ def __init__(
         self._watchers: list[Callable] = []
         self._fly_status: WatchableAsyncStatus | None = None
         self._fly_start: float
-        self._iterations_completed: int = 0
+        self._frames_collected: int = 0
+        self._current_number_index: int = 0
```

Comment:
We should call this `_current_[whatever_the_variable_in_trigger_is_called]_index`

```suggestion
        self._current_number_of_triggers_index: int = 0
```

Reply:
Actually, I think `_current_number_index` would be much better suited to be an iterator:
```suggestion
        self._number_of_triggers_iter: Iterator[int] | None = None
```
Then you can define it when you create the `trigger_info` as

``` python
    iter(
        trigger_info.number_of_triggers
        if isinstance(trigger.number_of_triggers)
        else iter([number_of_triggers])
    )
```

Then

``` python
        if isinstance(self._trigger_info.number, list):
            self._frames_collected += self._trigger_info.number[
                self._current_number_index
            ]
        else:
            self._frames_collected += self._trigger_info.number
```

becomes

``` python
    self.frames_to_complete = next(self.number_of_triggers_iter)
    self._frames_collected += self.frames_to_complete
```

and 

``` python
            if index >= self._trigger_info.number[self._current_number_index]:
                self._current_number_index += 1
                break
            else:
                if index >= self._trigger_info.number:
                    break
```

becomes

``` python
    if index == self._frames_to_complete:
        break
```